File: Python3.x/217-Contains Duplicate.py
# 暴力
# 先排序再遍历搜索
# Runtime: 48 ms, faster than 61.14% of Python3 online submissions for Contains Duplicate.
# Memory Usage: 16.4 MB, less than 90.16% of Python3 online submissions for Contains Duplicate.
class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        nums.sort()
        for i in range(1, len(nums)):
            if nums[i - 1] == nums[i]: return True
        return False


# set
# Runtime: 40 ms, faster than 95.12% of Python3 online submissions for Contains Duplicate.
# Memory Usage: 19 MB, less than 39.66% of Python3 online submissions for Contains Duplicate.
class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        return False if len(list(set(nums))) == len(nums) else True


# Time Limit Exceeded
class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        len_nums = len(nums)
        for i in range(len_nums):
            for j in range(i + 1, len_nums):
                if nums[i] == nums[j]: return True
        return False

File: C++/217-Contains Duplicate.cpp
// 暴力
// 先排序再遍历搜索
// Runtime: 24 ms, faster than 99.23% of C++ online submissions for Contains Duplicate.
// Memory Usage: 11.5 MB, less than 78.15% of C++ online submissions for Contains Duplicate.
class Solution {
public:
    bool containsDuplicate(vector<int>& nums) {
        sort(begin(nums), end(nums));
        for (int i = 1; i < nums.size(); ++i) {
            if (nums[i] == nums[i - 1])
                return true;
        }
        return false;
    }
};


// set
class Solution {
public:
    bool containsDuplicate(vector<int>& nums) {
        set<int> tmp;
        for (int i = 0; i < nums.size(); ++i) {
            if (tmp.find(nums[i]) != tmp.end())
                return true;
            else
                tmp.insert(nums[i]);
        }
        return false;
    }
};


// set
// Runtime: 48 ms, faster than 40.46% of C++ online submissions for Contains Duplicate.
// Memory Usage: 18.3 MB, less than 5.03% of C++ online submissions for Contains Duplicate.
class Solution {
public:
    bool containsDuplicate(vector<int>& nums) {
        return nums.size() > set<int>(nums.begin(), nums.end()).size();
    }
};


// set
// Runtime: 40 ms, faster than 76.39% of C++ online submissions for Contains Duplicate.
// Memory Usage: 16.7 MB, less than 18.05% of C++ online submissions for Contains Duplicate.
class Solution {
public:
    bool containsDuplicate(vector<int>& nums) {
        unordered_map<int, int> mp;
        for (int i : nums) {
            if (++mp[i] > 1)
                return true;
        }
        return false;
    }
};
